### ISSUE TITLE
Custom resources can't be saved when edited by plugin

### ISSUE THREAD

**Issue by vnen** [godotengine/godot#6603]
**Operating system or device - Godot version:**
Windows 10 64-bits. Recent master build.

**Issue description** (what happened, and what was expected):
I'm creating an editor for my custom resource, but when trying to use `ResourceSave.save()` with the object received in `edit()` the resource is not saved and there's an error in console:

```
ERROR: _ResourceSaver::save: Condition ' p_resource.is_null() ' is true. returned: ERR_INVALID_PARAMETER
   At: core\bind\core_bind.cpp:134
```

Saving the resource from the inspector works.

Also, testing for `object extends Resource` returns `true`, so it is actually is a Resource, but the reference is lost somehow in the process, so the `save()` method receives a null reference and gives the error.

**Steps to reproduce:**
- Create a plugin with a custom resource.
- Create a new resource of that type from the inspector.
- Try to save the custom resource using `ResourceSaver.save()` with the object received from `edit()` method.
- See errors in console and file not saved.

**Link to minimal example project** (optional but very welcome):
[custom-resource.zip](https://github.com/godotengine/godot/files/491475/custom-resource.zip)

This project has a sample plugin already enabled. Create a new `CustomResource` from the Inspector then click on the **Save** button on the top-right toolbar. There'll be an OS alert and the error will show in console.

**Comment by vnen** [comment]
My ugly workaround for now is to use `inspect_object()` and save through the Inspector.

**Comment by akien-mga** [comment]
See #9485 for a proposed fix (which was eventually rejected) and more discussion about potential fixes. The current status is that we're not sure how to fix it properly without impacting the performance too much.

**Comment by gtsiam** [comment]
I have posted about a different way to fix the problem in #9485, which will only affect performance when calling functions containing Ref arguments through the script system, by resolving whether or not typecasting is needed during the compile time. If you are interested, see my explanation in the pr (I don't know if I should create a new pr for this, so if I should, please tell me).
